### DISCUSSION TITLE
Using AUTH with X509Token on jgroups 5.2.2 runs into a NullPointerException

### DISCUSSION THREAD

**Discussion by joao-goncalves-jemmic** [belaban/JGroups#641] [Q&A] [answered]
Hello,

We have a use case where we are using jgroups with certificate based authentication (AUTH + X509Token). After upgrading to version 5.2.2 we started to have null pointer exceptions coming from the X509Token.setCertificate() because the keystore_path attribute was null. After checking out the code for version 5.2.2, adding a few log statements and using the modified library on our project, we could also see that all the other X509Token attributes, which were supposed to be initialized, were also null (screenshot below):
![image](https://user-images.githubusercontent.com/87442263/180470828-0dce739f-a766-4d53-bbe8-00d4f1ab4a2a.png)

This was the configuration which we are using (screenshot below  for better visualization):
`
<config xmlns="urn:org:jgroups" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:org:jgroups http://www.jgroups.org/schema/jgroups.xsd">
    
    <TCP
        bind_port="7800"
        bind_addr="localhost"
        thread_pool.min_threads="2"
        thread_pool.max_threads="8"
        thread_pool.keep_alive_time="30000"
    />
    <RED/>
    <TCPPING
    	async_discovery="true"
    	initial_hosts="localhost[7800],localhost[7801]"
    	port_range="0" />
    <MERGE3 max_interval="30000" min_interval="10000"/>
    <FD_SOCK2/>
    <FD_ALL3 timeout="40000" interval="5000"/>
    <VERIFY_SUSPECT2 timeout="1500" />
    <BARRIER/>
    <SYM_ENCRYPT
    	sym_algorithm="AES/CBC/PKCS5Padding"
        sym_iv_length="16"
        keystore_type="JCEKS"
        keystore_name="<path_to_keystore>/tcp_jgroups.keystore"
        store_password="changeit"
        alias="myKey"/>
    <pbcast.NAKACK2 use_mcast_xmit="false" discard_delivered_msgs="true"/>
    <UNICAST3 />
    <pbcast.STABLE desired_avg_gossip="50000" max_bytes="4M"/>
    <AUTH auth_class="org.jgroups.auth.X509Token"
        auth_token.auth_value="changeit"
        auth_token.cert_alias="jgroups_auth"
        auth_token.keystore_path="<path_to_keystore>/jgroups_auth.keystore"
        auth_token.keystore_password="changeit"
        auth_token.cipher_type="RSA"
    />
    <pbcast.GMS print_local_addr="true" join_timeout="2000"/>
    <UFC max_credits="2M" min_threshold="0.4"/>
    <MFC max_credits="2M" min_threshold="0.4"/>
    <FRAG2 frag_size="60K" />
    <pbcast.STATE_TRANSFER/>
</config>
`
![image](https://user-images.githubusercontent.com/87442263/180471843-37c653a9-cc5c-4bca-be15-7ae09ba6b404.png)

After a bit of investigation and digging into the code, it seems that for the specific X509Token case, the AUTH protocol has an inconsistency. 
When processing and initializing the AUTH protocol, the AUTH.init() method will be called inside ProtocolStack.initProtocolStack() on line 794:
![image](https://user-images.githubusercontent.com/87442263/180473317-6b4d802a-370e-497c-9bfa-126bcfe4ecbf.png)

The _AUTH.init()_, in case the _auth_token_ is of type _X509Token_, will then call the _X509Token.setCertificate()_ which expects the X509Token attributes to be initialized already.
![image](https://user-images.githubusercontent.com/87442263/180473949-3b2d7c0c-64f0-419e-9490-5d5dbf06ad47.png)

The issue is that the initialization of the X509Token attributes only happens after the processing of the Component annotation, which is taking place once the init() method has finished, _ProtocolStack.initProtocolStack()_ on line 795:
![image](https://user-images.githubusercontent.com/87442263/180474472-e6e8b2cd-ebe1-42d1-8987-6cb0462d69f1.png)

If we move the block of code which is calling the _X509Token.setCertificate()_ from _AUTH.init()_ to the beginning of _AUTH.start()_, the initialization and connection to the cluster are made successfully:
![image](https://user-images.githubusercontent.com/87442263/180475125-f668d1c8-bcf7-4d3a-bbd2-d6862742444b.png)
![image](https://user-images.githubusercontent.com/87442263/180475535-c87ae1ed-43cd-448d-9537-3896d69d5dfe.png)

We can create a fork of the repository and a PR for this and also to update the sample XML config files which are missing the _auth_token._ prefix on the X509Token related attributes but we are not sure if having the initialization of the X509Token on the start() method is ideal or not.

What would be your thoughts about this?

Cheers,

João

**Comment by belaban** [comment, marked as answer]
Hi Joao
yes, this looks like a good solution to me. Please create a JIRA first, then send the PR, so we have a record of the change.
Thanks!

**Comment by joao-goncalves-jemmic** [comment]
Hello,

Will do that. :)
Thanks.
